test/attn.py

```diff
@@ -0,0 +1,72 @@
+from torch import autograd
+import torch
+
+# Inherit from Function
+class AttentionFunction(autograd.Function):
+    @staticmethod
+    def attn(ctx, q, k, v):
+        x = torch.matmul(q, k.transpose(0, 1))
+        a = torch.tanh(x)
+        ctx.save_for_backward(q, k, v, x, a)
```

Comment:
since you saved `a` here you can replace usages of `tanh(x)` with that

Reply:
if you want your function to support double backward, you should only be saving the inputs and outputs your function. Because gradient tracking is disabled when forward is run, intermediate values computed during forward will not have gradients tracked. So, if we use them during the backward the contributions of the intermediary values to the outputs won't be properly added and the gradients would be wrong.

More context: https://pytorch.org/tutorials/intermediate/custom_function_double_backward_tutorial.html

Reply:
:o right!